src/Symfony/Component/Notifier/Message/PushMessage.php

```diff
@@ -16,7 +16,7 @@
 /**
  * @author Tomas Norkūnas <norkunas.tom@gmail.com>
  */
-final class PushMessage implements MessageInterface
+final class PushMessage implements PushMessageInterface
```

Comment:
I'd rather make PushMessage non final and remove the interface; value objects should not be final.

Reply:
that's what I think

Reply:
im willing to, but notice that said implementing an interface first introduce a contract and secondly allow implementation vs extension which its always good

Reply:
My experience is that interfaces are needed to abstract services, but here we are representing data. A value object, thus a class is SOLID enough to me.

Reply:
What's the use case? I made it final to be consistent with other messages. If you need something to extend you can do it through message options.

Reply:
The use case:

I'm implementing a transport for pusher for which I need to implement also a message which also leads me to implement a Notification which extends Notification and implements the PushNotificationInterface which has the method `public function asPushMessage(RecipientInterface $recipient, string $transport = null): ?PushMessage` fixed to PushMessage where the blocker is.

The current PushMessage is too strict in consideration to the recipient id so we need and want to give the flexibility to send to multiple pusher channels on the same call, another way will beat the purpose of the pusher API making several calls for the same data.

That said any of the above, an interface or taking the final out of the value object will allow developers to enhance with new transports

Reply:
And what about creating custom options object for pusher, where you could define multi recipients?

Reply:
i have not found on the doc or on the code anyway to pass options to the transport from the notifier service, I can only do it using a new notification and that blocks me because of the message as explained above

Reply:
`PushMessage` constructor accepts options as a third argument, so your transport could define options on it's own and then in the `asPushMessage` method you could do something like:

```
if ($transport === 'pusher') {
  $options = (new PusherOptions())->channels(..);

  return new PushMessage(.., $options);
}
```

And in the transport you could check if that option is defined then use it to make api call

Reply:
got it. I can close this now but I advise to look at the options to include interfaces or drop the final on value objects as Nicolas propose

Reply:
Putting a final on a class that is referenced by an interface in an API design bad practice to me, because it kills any extensibility. I'm still for removing the keyword.

Reply:
agreed and done